R/distribution.R

```diff
@@ -0,0 +1,203 @@
+#' Distribution is the abstract base class for probability distributions.
+Distribution <- R6::R6Class(
+  "torch_Distribution",
+  lock_objects = FALSE,
+  
+  public = list(
+    
+    has_rsample           = FALSE,
+    has_enumerate_support = FALSE,
+    `_validate_args`      = FALSE,
+    support               = NULL,
+    
+    # Choose different structure?
+    arg_constraints       = list(),
+    
+    initialize = function(batch_shape, event_shape, validate_args = NULL){
+      
+      self$batch_shape <- batch_shape
+      self$event_shape <- event_shape
+      
+      if (!is.null(validate_args))
+        self$validate_args <- validate_args
+      
+        for (p in seq_along(self$arg_constraints)) {
+          
+          constraint <- arg_constraints[[p]]$constraint
+          param      <- arg_constraints[[p]]$param
+          
+          if (constr_is_dependent(constraint))
+            next
+          if (!(param %in% as.list(self)) && inherits(param, "lazy_property"))
+            next # skip checking lazily-constructed args
+          if (all(constr_check(param)))
+            value_error("The parameter {param} has invalid values")
+        }
+    },
+    
+    expand = function(batch_shape, .instance = NULL){
+      stop("torch_Distribution is an abstract class.")
+    },
+    
+    #' Generates a sample_shape shaped sample or sample_shape shaped batch of
+    #' samples if the distribution parameters are batched.
+    sample = function(sample_shape=torch.Size()){
+      with_no_grad({
+        self$rsample(sample_shape)
+      })
+    },
+    
+    #' Generates a sample_shape shaped reparameterized sample or sample_shape
+    #' shaped batch of reparameterized samples if the distribution parameters
+    #' are batched.
+    #' TODO: torch_Size class?
```

Comment:
I believe it isn't. I asked this question in comment, because I'm curious if PyTorch developers had any special reason to use such class instead of simple integer vector.